invisible_cities/reco/paolina_functions.py

```diff
@@ -222,60 +223,60 @@ def make_tracks(evt_number       : float,
     return tc
 
 
-def drop_voxel(voxels: Sequence, the_vox: Voxel):
-    """Eliminate an individual voxel from a set of voxels and give its energy to the hit
-       that is closest to the barycenter of the eliminated voxel hits, provided that it
-       belongs to a neighbour voxel."""
-    ### remove voxel from list of voxels
-    try:
-        voxels.remove(the_vox)
-    except ValueError:
-        raise VoxelNotInList
-
-    if len(voxels) == 0:
-        return
-
-    pos = [h.pos for h in the_vox.hits]
-    qs = [h.E for h in the_vox.hits]
-    bary_pos = np.average(pos, weights=qs, axis=0)
-
-    ### find hit with minimum distance, only among neighbours
-    min_dist = 1e+06
-    min_hit = voxels[0].hits[0]
-    min_v = voxels[0]
-    for v in voxels:
-        if neighbours(the_vox, v):
-            for hh in v.hits:
-                dist = np.linalg.norm(bary_pos - hh.pos)
-                if dist < min_dist:
-                    min_dist = dist
-                    min_hit = hh
-                    min_v = v
-
-    ### add voxel energy to hit and to voxel, separately
-    min_hit.energy += the_vox.E
-    min_v.energy += the_vox.E
-
-
-def drop_end_point_voxels(voxels: Sequence[Voxel], energy_threshold: float, min_vxls: int):
+def drop_end_point_voxels(voxels: Sequence[Voxel], energy_threshold: float, min_vxls: int = 3) -> Sequence[Voxel]:
     """Eliminate voxels at the end-points of a track, recursively,
        if their energy is lower than a threshold"""
+
+    def drop_voxel(voxels: Sequence[Voxel], the_vox: Voxel):
+        """Eliminate an individual voxel from a set of voxels and give its energy to the hit
+           that is closest to the barycenter of the eliminated voxel hits, provided that it
+           belongs to a neighbour voxel."""
+
+        ### remove voxel from list of voxels
+        voxels.remove(the_vox)
+
+        pos = [h.pos for h in the_vox.hits]
+        qs = [h.E for h in the_vox.hits]
+        bary_pos = np.average(pos, weights=qs, axis=0)
+
+        ### find hit with minimum distance, only among neighbours
+        min_dist = 1e+06
+        min_hit = voxels[0].hits[0]
+        min_v = voxels[0]
+        for v in voxels:
+            if neighbours(the_vox, v):
+                for hh in v.hits:
+                    dist = np.linalg.norm(bary_pos - hh.pos)
+                    if dist < min_dist:
+                        min_dist = dist
+                        min_hit = hh
+                        min_v = v
+
+        ### add voxel energy to hit and to voxel, separately
+        min_hit.energy += the_vox.E
+        min_v.energy += the_vox.E
+
+
+    mod_voxels = copy.deepcopy(voxels)
```

Comment:
Why does a shallow copy not suffice?

Reply:
Because voxel contains hits and she is changing energies of the hits, so shallow copy wont protect us from keeping the hits of the original voxel list the same. On the other hand I dont quite like this new style, I think that drop voxels should be a function as it was in the original proposal.

Reply:
I suspected it had something to do with mutating the voxels, but then I was surprised to find that no tests fail if a shallow copy is performed instead.

Please provide a test that demonstrates the need for `deepcopy`, otherwise someone in the future will remove it as superfluous. Although if *none* of our extensive tests spot the problem ...

Reply:
Yes, my suggestion is to assert that the sum of hit energies for all hits inside the track before and after dropping voxels. That fails with the shallow copy.
I would also suggest to make E (and position maybe too) of a voxel a property of voxel that is being calculated based on hits contained inside, that way the voxel energy will be correctly updated when the hits energies inside the voxel change.

Reply:
1. We can add a test that ensures that the original list of hits is not changed by the function.
2. The E property of a voxel was originally kept independent on the sum of the hit energies on purpose as a check that the voxelization algorithm (`voxelize_hits`) was performing correctly. I would like to keep this check, in case the voxelization algorithm is changed, but maybe there's another way to do that.

Reply:
> We can add a test that ensures that the original list of hits is not changed by the function.

We could do this by construction, passing a tuple rather than a list.

> The E property of a voxel was originally kept independent on the sum of the hit energies on purpose as a check that the voxelization algorithm (voxelize_hits) was performing correctly. I would like to keep this check, in case the voxelization algorithm is changed, but maybe there's another way to do that.

Noting prevents us from adding a second number, say E_voxel, E_sumHits

Reply:
I put the `drop_voxel` function inside the `drop_end_point_voxels`function, in order to keep using a mutable object. I think that this is a case where mutable objects are useful, since it is a recursive algorithm (it keeps dropping voxels until a condition is reached). The alternative is to make a lot of useless copies, I think. Since there is no current user case for the `drop_voxel` function alone, I think we can implement it this way, and if in the future somebody asks for it, we can change the code.

Reply:
As for the need of a `deepcopy`: I'm using it because we decided that functions in IC shouldn't have side effects, therefore we don't want the function to modify its argument. I will add a test to ensure that the function does not modify the input voxels, but I cannot figure out a test that demonstrates the need for not modifying the argument.

Reply:
The question was not why you copy, the question was why the copy needs to be *deep*. The answer is twofold:

1. The elements are mutable. For immutable elements you only ever need a shallow copy.

2. The elements are mutated. If they were never mutated, you could get away with a shallow copy.

> I put the `drop_voxel` function inside the `drop_end_point_voxels` function, in order to keep using a mutable object.

More precisely: in order not to violate the IC principle that functions should not modify their inputs. `drop_voxel` mutates its input, so by making it an *internal* implementation detail of `drop_end_point_voxels`, nobody can call it and be nastily surprised.

Let's introduce a bit of nomenclature: functions which mutate their inputs are called *destructive*, while ones which do not, are called *non-destructive*. So, we want IC functions to be non-destructive wherever pragmatically possible. But what if we do end up needing to write a destructive function?

I really like the Scheme approach to this: any function whose name ends in an exclamation mark (`!`) is destructive. Frequently one finds two different versions of the same function (for example `foo!` and `foo`) one destructive and one non-destructive. The former exists because it is more efficient than the latter, the latter exists because it is safer than the former. The universal and concise conventional meaning of `!` minimizes nasty surprises.

Unfortunately `!` is not a valid constituent character in Python, so we can't adopt that exact convention. But perhaps we should consider formalizing some such convention in IC.

Anyway, I think that making `drop_voxel` internal to `drop_end_point_voxels` is a decent idea. If we discover the need to pull it out, we should use whatever convention we agree on, to make it clear that it is a destructive function.

Reply:
> Julia copies the same notation from Scheme! 

Unfortunately Julia is repeatedly failing to copy the similar `?` notation for predicates.

`_D` is probably the best we're going to come up with in Python. It's not bad: not too long but slightly ugly, and the ugliness is a good thing, as we want to feel a little discomfort when using destructive functions.